pubsub/pubsub.go

```diff
@@ -562,20 +562,37 @@ func (s *Subscription) Receive(ctx context.Context) (_ *Message, err error) {
 			// log.Printf("BATCH SIZE %d", batchSize)
 
 			go func() {
+				defer func() {
+					close(s.waitc)
+					s.waitc = nil
```

Comment:
I'm a bit concerned that previously this happened with the lock, and now it doesn't.

If I'm following the code correctly, I think the only way out of this infinite `for` loop is when the channel is closed, so can't we put this cleanup code there and make sure we have the lock when we do it?

Reply:
In fact, tests failed with a DATA RACE due to this:

WARNING: DATA RACE
Read at 0x00c000482088 by goroutine 20:
  gocloud.dev/pubsub.(*Subscription).Receive()
      /home/runner/work/go-cloud/go-cloud/pubsub/pubsub.go:554 +0x312
  gocloud.dev/samples/order.(*processor).handleRequest()
      /home/runner/work/go-cloud/go-cloud/samples/order/processor.go:63 +0xc7
  gocloud.dev/samples/order.TestHandleRequest()
      /home/runner/work/go-cloud/go-cloud/samples/order/processor_test.go:47 +0x52a
  testing.tRunner()
      /opt/hostedtoolcache/go/1.22.5/x64/src/testing/testing.go:1689 +0x21e
  testing.(*T).Run.gowrap1()
      /opt/hostedtoolcache/go/1.22.5/x64/src/testing/testing.go:1742 +0x44

Previous write at 0x00c000482088 by goroutine 22:
  gocloud.dev/pubsub.(*Subscription).Receive.func2.1()
      /home/runner/work/go-cloud/go-cloud/pubsub/pubsub.go:567 +0x64
  runtime.deferreturn()
      /opt/hostedtoolcache/go/1.22.5/x64/src/runtime/panic.go:602 +0x5d

Goroutine 20 (running) created at:
  testing.(*T).Run()
      /opt/hostedtoolcache/go/1.22.5/x64/src/testing/testing.go:1742 +0x825
  testing.runTests.func1()
      /opt/hostedtoolcache/go/1.22.5/x64/src/testing/testing.go:2161 +0x85
  testing.tRunner()
      /opt/hostedtoolcache/go/1.22.5/x64/src/testing/testing.go:1689 +0x21e
  testing.runTests()
      /opt/hostedtoolcache/go/1.22.5/x64/src/testing/testing.go:2159 +0x8be
  testing.(*M).Run()
      /opt/hostedtoolcache/go/1.22.5/x64/src/testing/testing.go:2027 +0xf17
  main.main()
      _testmain.go:89 +0x2e4

Goroutine 22 (finished) created at:
  gocloud.dev/pubsub.(*Subscription).Receive()
      /home/runner/work/go-cloud/go-cloud/pubsub/pubsub.go:564 +0x493
  gocloud.dev/samples/order.(*processor).handleRequest()
      /home/runner/work/go-cloud/go-cloud/samples/order/processor.go:63 +0xc7
  gocloud.dev/samples/order.TestHandleRequest()
      /home/runner/work/go-cloud/go-cloud/samples/order/processor_test.go:47 +0x52a
  testing.tRunner()
      /opt/hostedtoolcache/go/1.22.5/x64/src/testing/testing.go:1689 +0x21e
  testing.(*T).Run.gowrap1()
      /opt/hostedtoolcache/go/1.22.5/x64/src/testing/testing.go:1742 +0x44

Reply:
You are right about this. I concentrated on the msg return and missed that `waitc` was handled while in lock. WIll update that